STS groups : make group categories dynamic rather than fixed

guilpier-code opened the pull request:
We make groups of **S**hort **T**erm **S**torage dynamic. 
This means that accepted groups are no longer fixed in the code. Instead, any group can be defined by user.
It has consequences on the **Thematic Trimming**, which now accepts only one item (namely **STS BY GROUP**) regarding STS groups output variables.
Turning it to on prints all variables related to groups. 

Note : 
- We added a default group name (in case user doesn't specify a group name for a cluster) : "**sts-def-group**". Feel free to amend it.
- A new test (a study) was created. Please find it  [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/files/14802896/ST-dynamic-groups.zip). Look at its **user's notes** to know its intentions.
- upon a @sylvlecl suggestion, doc was updated accordingly

## Required
- [x] Merge https://github.com/AntaresSimulatorTeam/Antares_Simulator_Tests_NR/pull/18/files, generate new reference results

guilpier-code committed:
Change STS categories : first simplifications

guilpier-code committed:
Change STS categories : test - adding a dynamic output var printed in value and id files, and not in detail files.

guilpier-code committed:
Change STS categories : adding a dynamic output variable for STS injection, withdrawal and levels by group

flomnes commented on src/libs/antares/study/include/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -64,14 +64,16 @@ class Properties
 
-    static const std::map<std::string, enum Group> ST_STORAGE_PROPERTY_GROUP_ENUM;
 private:
+    static const std::map<std::string, enum Group> GROUP_TO_ENUM;
```
```suggestion
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/parts/short-term-storage/properties.cpp:
```diff
@@ -100,8 +101,10 @@ bool Properties::loadKey(const IniFile::Property* p)
-            it != Properties::ST_STORAGE_PROPERTY_GROUP_ENUM.end())
+        this->groupName = p->value.c_str();
+
+        if (auto it = Properties::GROUP_TO_ENUM.find(p->value.c_str());
+            it != Properties::GROUP_TO_ENUM.end())
         {
             this->group = it->second;
```
```suggestion
return true;
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/parts/short-term-storage/properties.cpp:
```diff
@@ -121,7 +124,10 @@ void Properties::save(IniFile& ini) const
+    // gp : we'll have to print simply (instead of subsequent already existing lines) :
+    // s->add("group", this->groupName);
+
+    for (const auto& [key, value] : GROUP_TO_ENUM)
```
```suggestion
```

guilpier-code replied:
Done

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/all.h:
```diff
@@ -148,10 +151,13 @@ typedef                           // Prices
-                          STstorageWithdrawalByCluster<STstorageLevelsByCluster<
-                          STstorageCashFlowByCluster<
-                            UnsupliedEnergy           // Unsuplied Energy
+                        // <ShortTermStorageByGroup
```
```suggestion
                         <ShortTermStorageByGroup
```

guilpier-code replied:
Done

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/varDynamic.h:
Remove

guilpier-code replied:
Done

guilpier-code committed:
Change STS categories : removing code related to fixed groups + removing old STS group output var

guilpier-code committed:
Change STS categories : avoid code duplication

guilpier-code committed:
Change STS categories : remove useless code

guilpier-code committed:
Change STS categories : remove useless code

guilpier-code commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,379 @@
+        }
+    }
+
+    void buildSurveyReport(SurveyResults& results,
```
**buildSurveyReport** is called when building the columns related to the current output variable in **synthesis** results.
If **buildSurveyReport** method is defined in an output var class, it is called.
If not, defined, it is the **IVariable** class's  **buildSurveyReport** that is called.
This function was a great source of confusion during this dev.
So I have 2 remarks about it : 
- We **really** **really** should rename it, because it has an horrible name, unrelated to what it does.
- take a look at **IVariable**'s version of this function : it calls **info.h**'s **BuildSurveyReport** function, and this latter function calls function **setClusterCaption** which is such an horrible function containing many **if**s. We should define a **buildSurveyReport** in variable classes for which **categoryFileLevel** is **Category::de_res** or **Category::de_sts**, so we could remove these **if**s. Actually this would allow to remove the horrible **setClusterCaption**, and make enclosing **info.h** : **BuildSurveyReport** to be the default behavior.

sylvlecl replied:
Seems to make sense!
Let's not do it in that PR though :)

guilpier-code commented:
CI failure :
The CI fails because some of our studies are not valid anymore.
Indeed, some of their data related **Thematic Trimming** are no more supported.
These studies for example specify **PSP_OPEN_INJECTION** as an output variable to be printed, and possibly 26 more keywords related to short term storage groups.
But now we allow only one item for all STS groups output variables to be printed  : **STS BY GROUP**.
So we need to update these failing studies' thematic trimming input.

guilpier-code committed:
Change STS categories : default value for default group name (in case user did not define a group for a STS cluster)

sylvlecl commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,379 @@
+
```
Copyright header is missing

guilpier-code replied:
Done

sylvlecl commented on src/tests/end-to-end/simple_study/simple-study.cpp:
```diff
@@ -318,7 +318,7 @@ BOOST_AUTO_TEST_CASE(sts_initial_level)
     props.efficiencyFactor = .9;
     props.initialLevel = .443;
-    props.group = Group::PSP_open;
+	props.groupName = std::string("Some STS group");
```
Wrong indent

guilpier-code replied:
Not in my **IDE** ! 
It seems like when visualing the code on **gitHub**, a tab character is not equivalent to 4 spaces, but larger.

guilpier-code replied:
Corrected

sylvlecl commented on src/libs/antares/study/include/antares/study/parts/short-term-storage/properties.h:
```diff
@@ -62,15 +62,14 @@ class Properties
-    /// Used to sort outputs
-    Group group = Group::Other1;
+    // Used to sort outputs
+    std::string groupName = "sts-def-group";
```
Question to be tackled:
should we keep Other1 here instead, for backwards compatibility?

flomnes replied:
```suggestion
    std::string groupName = "Other 1";
```

sylvlecl commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,379 @@
+        for (auto& cluster : area->shortTermStorage.storagesByIndex)
+        {
+            std::string groupName = cluster.properties.groupName;
+            if (std::find(groupNames_.begin(), groupNames_.end(), groupName) == groupNames_.end())
```
Note that a `std::set` would have given you name unicity and sorting without this and the std::sort at the end (replaced by a copy).
(and with more efficiency than a linear search in a vector, but here it does not matter since vectors are very small)
Something like
```cpp
std::set<std::string> names;
for (auto& cluster : area->shortTermStorage.storagesByIndex)
{
    names.insert(cluster.properties.groupName);
}
groupNames_ = std::vector<>(names.begin(), names.end());
```

guilpier-code replied:
Yes, I tried the **std::set** instead : it seemed indeed a good way to have code simpler.
But at some point (see function `std::string caption(unsigned int column) const`)  we need to get a variable name from a column number, and a **std::vector** is adapted in this situation.
So we still could have a **std::set** instead, but we may need another container for associating a string to a number.
Of course, I could miss something. Tell me if there's better to do.

guilpier-code replied:
ok, I understand your point better.
I made the simplifications you asked.

sylvlecl commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,379 @@
+
+    inline uint64_t memoryUsage() const
+    {
+        uint64_t r = (sizeof(IntermediateValues) * nbGroups_ + IntermediateValues::MemoryUsage())
```
I guess we should use nbColumns_ in this method instead?

guilpier-code replied:
Good eyes !
Corrected.
Even if we should remove this RAM usage, useless because only used in **GUI**, hopefully soon removed.
This memory usage consideration can be found in several places in the code.
Not sure removing it would be expensive, and many lines would vanish.
To me, it would worth it.

sylvlecl commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,379 @@
+}; // class VCard
+
+/*!
+** \brief Energy generated by short term storage clusters
```
Variables related to short term storage groups ?

guilpier-code replied:
Done

sylvlecl reviewed (commented):
A few minor comments

guilpier-code committed:
Change STS categories : correction due to review

guilpier-code committed:
Change STS categories : correction due to review

guilpier-code committed:
[skip ci] Change STS categories : add a license banner

guilpier-code committed:
[skip ci] Change STS categories : small simplifications

flomnes commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -778,6 +778,7 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
     area.spreadSpilledEnergyCost = 0.;
 
     bool ret = true;
+    auto studyVersion = study.header.version;
```
```suggestion
    const auto studyVersion = study.header.version;
```

guilpier-code committed:
[skip ci] Change STS categories : small clarification

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,391 @@
+    size_t nbColumns_ = 0;
+    std::vector<std::string> groupNames_; // Names of group containing the clusters of the area
+    std::map<std::string, unsigned int> groupToNumbers_; // Gives to each group (of area) a number
+    const std::vector<std::string> VAR_POSSIBLE_KINDS = { "injection", "withdrawal", "level" };
```
```suggestion
```

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,391 @@
+
+    std::string caption(unsigned int column) const
+    {
+        std::string groupName = groupNames_[column / NB_COLS_PER_GROUP];
+        std::string variableKind = VAR_POSSIBLE_KINDS[column % NB_COLS_PER_GROUP];
+        return groupName + "_" + variableKind;
```
```suggestion
        static const std::vector<std::string> VAR_POSSIBLE_KINDS = { "injection", "withdrawal", "level" };
        std::string groupName = groupNames_[column / NB_COLS_PER_GROUP];
        std::string variableKind = VAR_POSSIBLE_KINDS[column % NB_COLS_PER_GROUP];
        return groupName + "_" + variableKind;
```

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,391 @@
+        pNbYearsParallel = study->maxNbYearsInParallel;
+        pValuesForTheCurrentYear = new VCardType::IntermediateValuesBaseType[pNbYearsParallel];
+
+        // Building the vector of group names the clusters belong to.
+        std::set<std::string> tmp_names;
+        for (auto& cluster : area->shortTermStorage.storagesByIndex)
+            tmp_names.insert(cluster.properties.groupName);
+        groupNames_ = { tmp_names.begin(), tmp_names.end() };
```
Create a function for that

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,391 @@
+            tmp_names.insert(cluster.properties.groupName);
+        groupNames_ = { tmp_names.begin(), tmp_names.end() };
+        
+        // Giving a number to each group
+        unsigned int groupNumber{0};
+        for (auto name : groupNames_)
+        {
+            groupToNumbers_[name] = groupNumber;
+            groupNumber++;
+        }
```
Create a function for that

flomnes reviewed (commented):
Create functions, this will
- Not pollute the scope with temporary variables
- Allow unit-testing

guilpier-code committed:
[skip ci] Change STS categories : update documentation

flomnes committed:
Fix style for STS by group (#2011)

- Remove magic values (0, 1, 2), use `enum`
- Add a few `const` qualifiers
- Add const reference to results to improve readibility
- Formatting

flomnes commented on src/solver/variable/include/antares/solver/variable/adequacy/all.h:
```diff
@@ -210,8 +210,6 @@ typedef // Prices
                                           Variable::Economy::WaterValue,
                                           Common::SpatialAggregate<
                                             Variable::Economy::HydroCost,
```
What about districts ?

flomnes committed:
Add deprecation notice for STS legacy variables (#2016)

If a study has existing groups, emit a warning.

If the user does not `--force`, this warning is fatal. If the user uses
`--force`, the property is ignored and the simulation is launched.

guilpier-code committed:
STS groups : output titles to uppercase (#2017)

Any STS group name defined by the user is converted to uppercase in
output results.

flomnes committed:
Update src/libs/antares/study/area/list.cpp

flomnes committed:
Document changes for v9.1.0 (#2018)

Remove some unused code, too.

flomnes commented on src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h:
```diff
@@ -0,0 +1,406 @@
+    {
+        static const std::vector<std::string> VAR_POSSIBLE_KINDS
+          = {"INJECTION", "WITHDRAWAL", "LEVEL"};
+        std::string groupName = groupNames_[column / NB_COLS_PER_GROUP];
+        std::string variableKind = VAR_POSSIBLE_KINDS[column % NB_COLS_PER_GROUP];
+        return groupName + "_" + variableKind;
```
```suggestion
        const std::string& groupName = groupNames_[column / NB_COLS_PER_GROUP];
        const std::string& variableKind = VAR_POSSIBLE_KINDS[column % NB_COLS_PER_GROUP];
        return groupName + "_" + variableKind;
```

flomnes committed:
Update src/solver/variable/include/antares/solver/variable/economy/STSbyGroup.h
